gobblin-runtime/src/main/java/org/apache/gobblin/runtime/api/MySQLMultiActiveLeaseArbiter.java

```diff
@@ -0,0 +1,389 @@
+      else if (leaseValidityStatus == 2) {
+        if (isWithinEpsilon) {
+          LOG.warn("Lease should not be out of date for the same trigger event since epsilon << linger for flowAction"
+                  + " {}, db eventTimestamp {}, db leaseAcquisitionTimestamp {}, linger {}", flowAction,
+              dbEventTimestamp, dbLeaseAcquisitionTimestamp, dbLinger);
+        }
+        // Use our event to acquire lease, check for previous db eventTimestamp and leaseAcquisitionTimestamp
+        ResultSet rs = withPreparedStatement(
+            String.format(CONDITIONALLY_ACQUIRE_LEASE_IF_MATCHING_ALL_COLS_STATEMENT, this.leaseArbiterTableName,
+                this.leaseArbiterTableName, this.constantsTableName),
+            updateStatement -> {
+              completeUpdatePreparedStatement(updateStatement, flowAction, eventTimeMillis, true,
+                  true, dbEventTimestamp, dbLeaseAcquisitionTimestamp);
+              return updateStatement.executeQuery();
+            }, true);
+        return handleResultFromAttemptedLeaseObtainment(rs, eventTimeMillis);
+      } // No longer leasing this event
+        // CASE 5: Same event, no longer leasing event in db: terminate
+        if (isWithinEpsilon) {
+          return new NoLongerLeasingStatus();
+        }
+        // CASE 6: Distinct event, no longer leasing event in db
+        // Use our event to acquire lease, check for previous db eventTimestamp and NULL leaseAcquisitionTimestamp
+        ResultSet rs = withPreparedStatement(
+            String.format(CONDITIONALLY_ACQUIRE_LEASE_IF_FINISHED_LEASING_STATEMENT, this.leaseArbiterTableName,
+                this.leaseArbiterTableName, this.constantsTableName),
+            updateStatement -> {
+              completeUpdatePreparedStatement(updateStatement, flowAction, eventTimeMillis, true,
+                  false, dbEventTimestamp, null);
+              return updateStatement.executeQuery();
+            }, true);
+        return handleResultFromAttemptedLeaseObtainment(rs, eventTimeMillis);
+    } catch (SQLException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  /**
+   * Attempt lease by insert or update following a read based on the condition the state of the table has not changed
+   * since the read. Parse the result to return the corresponding status based on successful insert/update or not.
+   * @param resultSet
+   * @param eventTimeMillis
+   * @return LeaseAttemptStatus
+   * @throws SQLException
+   * @throws IOException
+   */
+  protected LeaseAttemptStatus handleResultFromAttemptedLeaseObtainment(ResultSet resultSet, long eventTimeMillis)
+      throws SQLException, IOException {
+    if (!resultSet.next()) {
+      throw new IOException("Expected num rows and lease_acquisition_timestamp returned from query but received nothing");
+    }
+    int numRowsUpdated = resultSet.getInt(1);
+    long leaseAcquisitionTimeMillis = resultSet.getTimestamp(2).getTime();
```

Comment:
yes, but it will be safeguarded by use of `leaseValidityStatus` which will have value of `3` if `lease_acquisition_status` is null which case we will not attempt to use this value.